### PR TITLE
Added HTTP Range

### PR DESCRIPTION
**Pull request by vhvb1989**
Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/1136

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 1]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
...
+   * `Offset + Length - 1` inclusively.
+   *
+   */
+  struct Range
+  {
+    /**
+     * @brief The starting point of the HTTP Range.
+     *
+     */
+    int64_t Offset = 0;
```
Comment (CaseyCarter) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:189]: I think we're missing `#include <stdint.h>` for `int64_t`.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 2]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
     NetworkAuthenticationRequired = 511, ///< HTTP 511 Network Authentication Required.
   };
 
+  /**
+   * @brief Defines a range of bytes within an HTTP resource, starting at an `Offset` and ending at
+   * `Offset + Length - 1` inclusively.
+   *
+   */
+  struct Range
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:183]: I understand this is where we landed in the issue, but re-consider naming is `HttpRange`.

File: sdk/core/azure-core/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,7 @@
 ### New Features
 
 - Added a WinHTTP-based `HttpTransport` called `WinHttpTransport` and use that as the default `TransportPolicyOptions.Transport` on Windows when sending and receiving requests and responses over the wire.
+- Added `Range` type to `Http` namespace.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:8]:
```suggestion
- Added `Range` type to `Azure::Core::Http` namespace. 
```

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 4]
```diff
@@ -486,10 +510,10 @@ namespace Azure { namespace Core { namespace Http {
...
     explicit Request(HttpMethod httpMethod, Url url, bool downloadViaStream)
         : Request(
-            httpMethod,
-            std::move(url),
-            NullBodyStream::GetNullBodyStream(),
-            downloadViaStream)
+              httpMethod,
+              std::move(url),
+              NullBodyStream::GetNullBodyStream(),
+              downloadViaStream)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:516]: Revert this formatting change, not sure why we keep introducing it.

File: sdk/core/azure-core/src/http/http.cpp [hunk 5]
```diff
@@ -153,3 +155,13 @@ void Azure::Core::Http::Details::InsertHeaderWithValidation(
   // insert (override if duplicated)
   headers[headerName] = headerValue;
 }
+
+std::string Azure::Core::Http::Internal::RangeToString(Range range)
+{
+  std::string rangeString = "bytes=" + std::to_string(range.Offset) + "-";
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/http.cpp:161]: nit: use a named `unit` constant like this:
https://github.com/Azure/azure-sdk-for-net/blob/8b100f13b70aa00fdb69304500ea7e2a6a142322/sdk/core/Azure.Core/src/HttpRange.cs#L17-L18

File: sdk/core/azure-core/src/http/http.cpp [hunk 6]
```diff
@@ -153,3 +155,13 @@ void Azure::Core::Http::Details::InsertHeaderWithValidation(
...
+  if (range.Length != std::numeric_limits<decltype(range.Length)>::max())
```
Comment (CaseyCarter) [inline sdk/core/azure-core/src/http/http.cpp:162]: Why use `2^63 - 1` as the "until the end" sentinel value when `-1` (or any negative length, really) would be easier to handle?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 7]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
...
+     *
+     */
+    int64_t Offset = 0;
+
+    /**
+     * @brief The size of the HTTP Range.
+     *
+     */
+    int64_t Length = 0;
+  };
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:196]: Should we add a ctor and validate offset/length are not negative? And that length isn't 0?

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:196]: What about equality/comparison operators?

File: sdk/core/azure-core/src/http/http.cpp (hunk 5 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/src/http/http.cpp:161]: I don't think the format is correct.

it's actually [startOffset, startOffset + length - 1]

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 8]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
...
+
+  namespace Internal {
+    std::string RangeToString(Range range);
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: If you have `ToString` here, you'll have to add `FromString`. Or you can just remove both of them.

I prefer you remove both

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:183]: It's already in `Http` namespace.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 9]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
...
+     * @brief The starting point of the HTTP Range.
+     *
+     */
+    int64_t Offset = 0;
+
+    /**
+     * @brief The size of the HTTP Range.
+     *
+     */
+    int64_t Length = 0;
```
Comment (CaseyCarter) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:195]: Should `Length` instead default to the "gimme everything" sentinel value, so `Range{42}` (omitting the length) produces `bytes=42-` (also omitting the length)?

Should this class have a `static constexpr` member to name the "gimme everything" sentinel value?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: Why do we have to add `FromString`? Is there a need?

File: sdk/core/azure-core/src/http/http.cpp (hunk 6 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/http.cpp:162]: The length needs to be nullable, it's optional, and we should use the fact it isn't set as the sentinel to leave it out of the string.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 9 above)
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:195]: Make `Length` nullable.

Comment (CaseyCarter) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:195]: FWIW, I prefer this to my suggestions regarding the sentinel value.

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:195]: Yep, we do want `bytes=42-` when length is omitted, and not `bytes=42-0`. We can solve this by making length a `Nullable<int64_t>`, and we should do that :)

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:195]: +1 for nullable

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: Well, I guess we don't

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 10]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
...
+  };
+
+  namespace Internal {
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:198]: `Internal` instead of `Details`?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: > Why do we have to add FromString? Is there a need?

It is currently there on Storage code. That's why I added here.  The Range needs to become an HTTP header at the end. That's why it needs to become string.

@Jinming-Hu  will the Server ever return a Range header that we need to parse?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:189]: I prefer `#include <cstdint>`

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: Well, yes and not. Server returns something like `ContentRange: bytes=100-110/200". But storage team can parse in our own code.

**Comment by Jinming-Hu** [comment]
Perfectly suits current situation😄

image

**Comment by ahsonkhan** [comment]
I had a similar thought @Jinming-Hu, but in this case I think we all just want to help @vhvb1989 with this PR because of his impressive turnaround ;)

**Comment by Jinming-Hu** [comment]
> I had a similar thought @Jinming-Hu, but in this case I think we all just want to help @vhvb1989 with this PR because of his impressive turnaround ;)

Yes, absolutely. I also have a lot of comments in this PR. I just said it for fun.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 7 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:196]: > Should we add a ctor and validate offset/length are not negative? And that length isn't 0?

Another case for pre-condition check, which we will discuss in tomorrow's guideline meeting.

Should we throw or crash if the input is invalid.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:196]: > What about equality/comparison operators?

I cannot think of its being used anywhere. We can always add it later.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: @vhvb1989, to clarify I asked about `FromString`, **not** `ToString`.

> It is currently there on Storage code.

Where?

> That's why I added here. 

Where? I only see `RangeToString`.

> The Range needs to become an HTTP header at the end. That's why it needs to become string.

Right, `ToString` is fine, keep it, I was posing the question to Jinming that just because we have one (internal) method, doesn't mean we need to add its counter. :)

> Server returns something like `ContentRange: bytes=100-110/200". But storage team can parse in our own code.

From what I can gather content range and this range type we are adding are not exactly the same thing and don't have the same format necessarily either (no equal sign, for example `Content-Range: bytes 200-1000/67589`).

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 10 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:198]: I believe the intent is that it will be called by the Storage package. `Details` is for private to the package only.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 7 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:196]: > We can always add it later.

Sounds good to me.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: It makes the most sense to have ToString on the HttpRange class.  I don't see a good reason to not provide that method.

**Review by RickWinter — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 11]
```diff
@@ -175,6 +175,30 @@ namespace Azure { namespace Core { namespace Http {
     NetworkAuthenticationRequired = 511, ///< HTTP 511 Network Authentication Required.
   };
 
+  /**
+   * @brief Defines a range of bytes within an HTTP resource, starting at an `Offset` and ending at
+   * `Offset + Length - 1` inclusively.
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:180]: `-1`? So, if length is 0, the end is before offset? If length is 1, the range is empty? If length is 2, the range contains one element?

Regardless of `-1`, can range be negative? Can offset be negative? Can Offset point out of the bounds of the enclosing range? What if length ends up end of the range pointing outside of the enclosing range?

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:180]: The way .NET solved those questions is by not allowing negative length/offset, and only allowing length to be > 0, if it iis provided.
https://github.com/Azure/azure-sdk-for-net/blob/8b100f13b70aa00fdb69304500ea7e2a6a142322/sdk/core/Azure.Core/src/HttpRange.cs#L38-L41

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: I will remove toString.
It might end up to be different for each service

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:199]: It would be each Service the responsible for adding a toString implementation. Since the way the want to add the header could change.
If it always return "bytes=x-y" and in the future another service want a different unit, or things like that, we can add extra constructors, but toString() would be hard to update

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp]: There is no reason to remove ToString. It is useful to have one standard (most commonly used) string representation of HttpRange that most callers will use. Any service that wants to change the behavior and create there own can always create their own helper and write it differently. By adding one for `HttpRange` were aren't preventing them from doing that.
Consider seeing if any service in .NET overrides the HttpRange.ToString value and how common that is. We can pick one format that is common enough to work for N services, and let the rest override.

File: sdk/core/azure-core/src/http/http.cpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/http.cpp]: Won't provide the toString method

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp]: Given that the storage is the only one currently asking for this, I won't add the toString() for now.
Since this is not exposed to public API, we can decide later if we want to add it.

I think an end-user doesn't really need to ever call toString() on the Range.  It is only the Service the one who needs it for adding a header.
As an end-user, I could expect to get something like `range={offset: x, length: y}` for toString() because we are adapting `toString()` for the format as it is expected to be as part of an HTTP header.  We are also hardocding the `bytes` which is not cool.

So, makes more sense not to add it for now

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp]: Also, if we make it public, why would an end-user want to call toString() anyway?
This was supposed to be Internal only.

**Review by JeffreyRichter — APPROVED** [review]

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/src/http/http.cpp [hunk 12]
```diff
@@ -3,6 +3,8 @@
 
 #include "azure/core/http/http.hpp"
 
+#include <utility>
+
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/http.cpp:7]: Avoid modifying files like this that are unrelated to the PR.

**Comment by ahsonkhan** [comment]
Should we add a ctor and validate offset/length are not negative? And that length isn't 0?

**Comment by vhvb1989** [comment]
> 
> 
> Should we add a ctor and validate offset/length are not negative? And that length isn't 0?

I think that validation needs to happen on Server side. The SDK usually does not do input validation

**Comment by ahsonkhan** [comment]
But won't the SDK/protocol layer be creating instances of `Range`s? What if the client library authors write buggy code and create an invalid range with negative values before passing it to the service?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 1 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:191]: Should we put this in the guidelines? Is `cstdint` the preferred/idiomatic way of doing it in C++ and is there something wrong about using `stdint.h`? I saw some posts in SO favoring cstdint and leaving stdint.h for C programs. @CaseyCarter - worth being consistent on or does it largely not matter?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 4 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:514]: I would like to understand what formatting/text editor you are using because we keep making this change, and if I format the `http.hpp` file again in VS, it reverts it:
image

**Comment by ahsonkhan** [comment]
This isn't MQ related, everything else belongs in January (the next milestone we release v-next), even if we finished them early.
